docs/design/antispam.md

```diff
@@ -0,0 +1,138 @@
+guarantee about uniqueness of entries in a log. These properties are not generally required, but are
+expensive and have a profound impact on log performance.
+
+If your transparency application needs such properties you will need to handle this "upstream" of Tessera.
+
+## Overview
+
+The anti-spam support is optional, and needs to be explicitly provisioned in the infrastructure and enabled
+by the personality application (via the
+[`WithAntispam`](https://pkg.go.dev/github.com/transparency-dev/trillian-tessera@main#AppendOptions.WithAntispam))
+ option).
+
+The following diagram gives a high-level overview of how it works.
+
+Across the top, we see the chain of _decorators_ ultimately leading to the storage implementation's `Append` func.
+The anti-spam implementation is part of this chain, but only performs _lookups_ here, short-circuiting and returning
+a previously assigned index if match is found.
+
+The anti-spam index data is populated asynchronously, by "following" (or "tailing") the contents of the log, 
+and adding index information for new entries as they appear.
+
+```mermaid
+architecture-beta
+  group dedup[Antispam implementation]
+  group appender[Storage Appender]
+  group follower[Log Follower]
+
+  service app(server)[Application]
+  service inmem(server)[In memory dedup] in dedup
+  service persistent(server)[Persistent dedup] in dedup
+  service dedupdata(disk)[Persistent index] in dedup
+  service populater(server)[Persistent index populater] in dedup
+  service append(server)[Append] in appender
+  service integrate(server)[Integrate] in appender
+  service bundles(disk)[Entry bundles] in appender
+  junction f1 in follower
+  junction f2 in follower
+
+  app:R --> L:inmem
+  inmem:R --> L:persistent
+  persistent:R --> L:append
+  append:R --> L:integrate
+
+  persistent:B -[Lookup]-> T:dedupdata
+  populater:T -[Insert]-> B:dedupdata
+
+  integrate:B -[Write]-> T:bundles
+  bundles:B -[Read]- T:f2
+  f2:L -- R:f1
+  f1:L --> R:populater
+```
+
+If the process of following the log falls too far behind a configurable threshold, the decorator will start
+returning `ErrPushback` to the application until such time as the follower has caught up. This helps to prevent
+the anti-spam mechanism getting so far behind as to become ineffective in preventing abuse.
+
+## Threats / failure scenarios
+
+### Threats:
+
+1. Stuttering but non-malicious submission client
+2. Malicious bot (e.g. feeding log contents back into the log)
+
+### Failure scenarios:
+
+1. Normal operation
+2. Degraded anti-spam
+3. Degraded integration
+
+As currently implemented, the Tessera anti-spam implementation consists of a layered defense: 
+
+1. An in-memory deduplication cache, followed by:
+2. A persistent deduplication index.
+
+This gives fast and cheap protection against (T1) in all failure scenarios, and falls back to the more costly interaction
```

Comment:
You may have elided this for brevity, but in reality the in-memory version may allow N copies through, where N is the number of Tessera Appender instances you have (i.e. probably the number of personality instances).

Reply:
I did consciously leave it out, but it is important so I've added a note about that.